reference/src/abort-and-assert/clever-errors.md

```diff
@@ -0,0 +1,230 @@
+    }
+}
+
+module 0x42::user_module {
+    use 0x42::macro_exporter::{
+        assert_false,
+        abort_always,
+        assert_false_fun,
+        abort_always_fun
+    };
+
+    fun invoke_assert_false() {
+        assert_false!(); // Will abort with the line number of this invocation
+    }
+
+    fun invoke_abort_always() {
+        abort_always!(); // Will abort with the line number of this invocation
+    }
+
+    fun invoke_assert_false_fun() {
+        assert_false_fun(); // Will abort with the line number of the assertion in `assert_false_fun`
+    }
+
+    fun invoke_abort_always_fun() {
+        abort_always_fun(); // Will abort with the line number of the `abort` in `abort_always_fun`
+    }
+}
+```
+
+## Inflating Clever Abort Codes
+
+Precisely, the layout of a clever abort code is as follows:
+
+```
+
+|<tagbit>|<reserved>|<source line number>|<module identifier index>|<module constant index>|
++--------+----------+--------------------+-------------------------+-----------------------+
+| 1-bit  | 15-bits  |       16-bits      |     16-bits             |        16-bits        |
+
+```
+
+Note that the Move abort will come with some additional information -- importantly in our case the
+module where the error occurred. This is important because the identifier index, and constant index
+are relative to the module's identifier and constant tables (if not set the sentinel values).
+
+> To decode a clever abort code, you will need to know the module where the error occurred if either
+> the identifier index or constant index are not set to the sentinel value of `0xffff`.
+
+In pseudo-code, you can decode a clever abort code as follows:
+
+```rust
+// Information available in the MoveAbort
+let clever_abort_code: u64 = ...;
+let (package_id, module_name): (PackageStorageId, ModuleName) = ...;
+
+let is_clever_abort = (clever_abort_code & 0x8000_0000_0000_0000) != 0;
+
+if is_clever_abort {
+    // Get line number, identifier index, and constant index
+    // Identifier and constant index are sentinel values if set to '0xffff'
+    let line_number = ((clever_abort_code & 0x0000_ffff_0000_0000) >> 32) as u16;
+    let identifier_index = ((clever_abort_code & 0x0000_0000_ffff_0000) >> 16) as u16;
+    let constant_index = ((clever_abort_code & 0x0000_0000_0000_ffff)) as u16;
+
+    // Print the line error message
+    print!("Error from '{}::{}' (line {})", package_id, module_name, line_number);
+
+    // No need to print anything or load the module if both are sentinel values
+    if identifier_index == 0xffff && constant_index == 0xffff {
+        return;
+    }
+
+    // Only needed if constant name and value are not 0xffff
+    let module: CompiledModule = fetch_module(package_id, module_name);
+
+    // Print the constant name (if any)
+    if identifier_index != 0xffff {
+        let constant_name = get_identifier_at_table_index(module, identifier_index);
```

Comment:
> `get_identifier_at_table_index`

I appreciate this is pseudo code, but do we have a method for this?

Reply:
We do, but its name is not quite as descriptive as I would like here (`identifier_at`). I'll try and make it a bit closer, but I think having a more descriptive name here will be beneficial.